File: src/main/java/at/yawk/wm/dock/module/DockBootstrap.java
package at.yawk.wm.dock.module;

import at.yawk.wm.dock.Widget;
import at.yawk.wm.dock.module.feature.DockPadFeature;
import at.yawk.wm.dock.module.feature.ScrollTagChange;
import at.yawk.wm.dock.module.widget.*;
import java.util.Arrays;
import java.util.List;
import javax.inject.Inject;

/**
 * @author yawkat
 */
public class DockBootstrap {
    @Inject DockBuilder dock;

    @Inject BatteryWidget batteryWidget;
    @Inject ClockWidget clockWidget;
    @Inject CpuWidget cpuWidget;
    @Inject MediaWidget mediaWidget;
    @Inject MemoryWidget memoryWidget;
    @Inject NetworkWidget networkWidget;
    @Inject ProgressWidget progressWidget;
    @Inject TagListWidget tagListWidget;
    @Inject TitleWidget titleWidget;

    @Inject DockPadFeature dockPadFeature;
    @Inject ScrollTagChange scrollTagChange;

    public List<Widget> getWidgets() {
        return Arrays.asList(
                batteryWidget,
                clockWidget,
                cpuWidget,
                mediaWidget,
                memoryWidget,
                networkWidget,
                progressWidget,
                tagListWidget,
                titleWidget
        );
    }

    public List<Runnable> getDockStartListeners() {
        return Arrays.asList(
                dockPadFeature::init,
                scrollTagChange::listen
        );
    }

    public void startDock() {
        dock.start(this);
    }
}

File: src/main/java/at/yawk/wm/dock/module/DockBootstrap.kt
package at.yawk.wm.dock.module

import at.yawk.wm.PeriodBuilder
import at.yawk.wm.di.PerMonitor
import at.yawk.wm.dock.module.feature.DockPadFeature
import at.yawk.wm.dock.module.feature.ScrollTagChange
import at.yawk.wm.dock.module.widget.BatteryWidget
import at.yawk.wm.dock.module.widget.ClockWidget
import at.yawk.wm.dock.module.widget.CpuWidget
import at.yawk.wm.dock.module.widget.MediaWidget
import at.yawk.wm.dock.module.widget.MemoryWidget
import at.yawk.wm.dock.module.widget.NetworkWidget
import at.yawk.wm.dock.module.widget.ProgressWidget
import at.yawk.wm.dock.module.widget.TagListWidget
import at.yawk.wm.dock.module.widget.TitleWidget
import at.yawk.wm.ui.Widget
import javax.inject.Inject

@PerMonitor
class DockBootstrap @Inject constructor(
    val dock: DockBuilder,
    val periodBuilder: PeriodBuilder,

    batteryWidget: BatteryWidget,
    clockWidget: ClockWidget,
    cpuWidget: CpuWidget,
    mediaWidget: MediaWidget,
    memoryWidget: MemoryWidget,
    networkWidget: NetworkWidget,
    progressWidget: ProgressWidget,
    tagListWidget: TagListWidget,
    titleWidget: TitleWidget,
    private val dockPadFeature: DockPadFeature,
    private val scrollTagChange: ScrollTagChange
) {
    val widgets: List<Widget> = listOf(
        batteryWidget,
        clockWidget,
        cpuWidget,
        mediaWidget,
        memoryWidget,
        networkWidget,
        progressWidget,
        tagListWidget,
        titleWidget
    )

    val dockStartListeners = listOf(Runnable { dockPadFeature.init() }, Runnable { scrollTagChange.listen() })

    fun startDock() {
        dock.start(this)
    }
}